Add feature to impersonate/proxy user sessions

shiyer7474 opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):

- New Feature

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):

Added new SQL statement EXECUTE AS to support user impersonation. Resolves #39048

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

WIP

### Details

Add the ability to impersonate or proxy an end-user session over an application admin user connection.

Resolves #39048

shiyer7474 committed:
First pass implementation of IMPERSONATE

shiyer7474 committed:
Completed implementation and testing

shiyer7474 commented:
- PR does not handle the `max_sessions_for_user `setting. Need to figure out the background on why `SessionTracker::Session `instance is a a member of the `Session `class and can it be moved to the `Session Context` class.
- Alternatives to adding a new SQL (`EXECUTE AS`) are - 
1) Implement the feature by enhancing the client/server protocol to add a `proxied_user `field. Corresponding changes to drivers maybe required. 
OR
2) Just introduce a new special notation for the username to indicate that impersonation/proxy'ing is requested : `appuser[serviceuser].`  Implementation could be simple compared to above  2 approaches - Validate `appuser `& `serviceuser`, perform authentication as `serviceuser`, check `serviceuser `has impersonation privileges on `appuser`, setup new session context to `appuser`.

UnamedRus commented:
> Just introduce a new special notation for the username to indicate that impersonation/proxy'ing is requested : appuser[serviceuser]. 

Somewhat theoretical, but it could affect/change result of queries.
Imagine, if there is ROW POLICY set as `WHERE user = currentUser();` or `WHERE groups IN dictGet('groups_for_user', 'groups', currentUser());`

filimonov commented:
One issue with executing multiple statements in sequence (e.g., `EXECUTE AS`) is the requirement to preserve the session between them. This can present challenges in various scenarios where session management is not practical or adds unnecessary complexity.

Consider the following cases:

1. **Stateless HTTP Clients**: Many HTTP clients do not implement session logic, working without session IDs or timeouts. Introducing session management (session ID generation, timeouts, etc.) could be a significant barrier, potentially becoming a 'show-stopper' for clients designed to operate statelessly.

2. **Load Balancers**: In configurations with load balancers, requests might be routed to different servers within a cluster. Since session state typically lives on the server level (not the cluster level), this could disrupt the user context continuity across requests.

3. **Connection Pools**: When client-side connection pools are used, subsequent requests may utilize different physical connections. Expecting session-based user context preservation across those connections would be cumbersome and error-prone.

In light of these issues, we need a way to switch users "in one shot," either during connection establishment or within a single SQL statement, without depending on session persistence. Protocol extensions might increase complexity for driver authors, which could become a barrier to adoption.

### I would suggest considering three potential solutions:

1. **Special Username Format**: A simple solution could involve using a special format for the username, such as `serviceuser(as:appuser)`. This would allow the application user to be specified within the same string, making it trivial for client drivers to pass. The server would then handle the user switch seamlessly, without the need for session management.

2. **User Context as a Setting**: Another approach could be to pass the target user as a setting. This can be done by passing settings in the query headers via the native protocol ([example](https://github.com/ClickHouse/ClickHouse/blob/b0da2a5534d5e3325f77454e252f1ea475571ce3/src/Client/Connection.cpp#L800)) or as query string parameters in the HTTP protocol. This method requires minimal to no changes for driver maintainers—they would simply need to pass the setting in the connection string. However, this must be done securely to ensure that end users cannot easily manipulate these settings.

3. **One-Shot SQL Statement**: A third option would be to extend SQL with a one-shot `EXECUTE AS user <query>` statement. For example, `EXECUTE AS misha SELECT ...` would allow user context switching within a single query, eliminating the need for session preservation and simplifying the logic for client-side implementations.

shiyer7474 commented:
Thanks @filimonov for the inputs regarding "one shot" switch user. My preference is also option 1 - special format user name like `'serviceuser(as:appuser)'` or `'serviceuser[appuser]'`. The current PR will need to remove all the code added to support the new SQL "EXECUTE AS" and just add authentication time code like this : Validate `appuser `& `serviceuser`, perform authentication as `serviceuser`, check `serviceuser `has impersonation privileges on `appuser`, setup new session context as`appuser`.

shiyer7474 commented:
@vitlibar Thanks for picking up this  PR for review! Please review (and associated issue #39048) and let us know your opinion.

shiyer7474 commented:
@vitlibar Let us know your feedback after initial review!

shiyer7474 commented:
@vitlibar Let us know if you got a chance to review the approach and it aligns well, I can then mark the PR ready for review. Thanks!

vitlibar commented on src/Parsers/Access/ASTExecuteAsQuery.h:
```diff
@@ -0,0 +1,23 @@
+class ASTRolesOrUsersSet;
+
+/** EXECUTE AS <user>
+  */
```
It would be nice to have a command to execute only one statement as a specified user, then get back to normal mode. Perhaps something like
```
EXECUTE AS <user> SELECT ...
```

vitlibar replied:
It may be also useful to enable switching this feature off:
```
EXECUTE AS <user> ON
...
EXECUTE AS <user> OFF
```

Simple `Execute AS <user>` could be made an alias of `EXECUTE AS <user> ON`.

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -1561,6 +1561,14 @@ void Context::setCurrentRolesDefault()
+    auto new_user_uuid = *(users.ids.begin());
+    setUser(new_user_uuid);
+    setCurrentUserName(getUser()->getName());
+    setInitialUserName(getUser()->getName());
```
Why do we need to set the initial user name?

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -5150,6 +5158,12 @@ void Context::setInitialUserName(const String & initial_user_name)
     need_recalculate_access = true;
 }
 
+void Context::setAuthUserName(const String & auth_user_name)
```
Do we call it somewhere?

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -1561,6 +1561,14 @@ void Context::setCurrentRolesDefault()
     setCurrentRolesImpl(user->granted_roles.findGranted(user->default_roles), /* throw_if_not_granted= */ false, /* skip_if_not_granted= */ false, user);
 }
 
+void Context::switchImpersonateUser(const RolesOrUsersSet & users )
```
It seems this function is used only once and it's very simple. Let's move it to `InterpreterExecuteAs.cpp` perhaps?

vitlibar commented on src/Interpreters/Access/InterpreterExecuteAsQuery.h:
```diff
@@ -0,0 +1,26 @@
+struct RolesOrUsersSet;
+struct User;
+
+class InterpreterExecuteAsQuery : public IInterpreter, WithMutableContext
```
`InterpreterExecuteAs` looks a bit better (because it's rather `execute as <user>`, not `execute as <query>`, and `InterpreterExecuteAsUserQuery` is just a bit too long).

vitlibar commented on src/Core/ServerSettings.h:
```diff
@@ -176,6 +176,7 @@ namespace DB
     M(Bool, disable_insertion_and_mutation, false, "Disable all insert/alter/delete queries. This setting will be enabled if someone needs read-only nodes to prevent insertion and mutation affect reading performance.", 0) \
     M(UInt64, keeper_multiread_batch_size, 10'000, "Maximum size of batch for MultiRead request to [Zoo]Keeper that support batching. If set to 0, batching is disabled. Available only in ClickHouse Cloud.", 0) \
     M(Bool, use_legacy_mongodb_integration, true, "Use the legacy MongoDB integration implementation. Note: it's highly recommended to set this option to false, since legacy implementation will be removed in the future. Please submit any issues you encounter with the new implementation.", 0) \
+    M(Bool, allow_impersonate_user, true, "Enable/disable the IMPERSONATE feature (EXECUTE AS <user>).", 0) \
```
We usually put our experimental feature flags in Core/Settings.cpp, it seems it's better to put this flag there too.

vitlibar replied:
Let's rename it to `allow_execute_as_user` (because `execute_as` is just easier to type :) )

vitlibar commented on tests/queries/0_stateless/03252_execute_as.config.xml:
```diff
@@ -0,0 +1,22 @@
+	<path>./</path>
+        </local_directory>
+    </user_directories>
+</clickhouse>
```
This config file won't be needed if you move `allow_impersonate_user` to `Core/Settings.cpp`.

vitlibar commented:
@shiyer7474 Sorry for the waiting.
This is a cool PR, I like it!

shiyer7474 commented:
@vitlibar Thanks for the detailed review! Please give me 2 days to incorporate and test the changes.

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #70775 from shiyer7474/ClickHouse:impersonate_user -->
*This is an automated comment for commit c11eb8e31de1081cf81bc4c6248e017198bb2c54 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70775/c11eb8e31de1081cf81bc4c6248e017198bb2c54/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/74454/64a78170b7f6ed84fd887fede8cce9cc3e519fdc/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

shiyer7474 commented on src/Access/Common/AccessType.h:
```diff
@@ -147,6 +147,7 @@ enum class AccessType : uint8_t
     M(SHOW_QUOTAS, "SHOW CREATE QUOTA", GLOBAL, SHOW_ACCESS) \
     M(SHOW_SETTINGS_PROFILES, "SHOW PROFILES, SHOW CREATE SETTINGS PROFILE, SHOW CREATE PROFILE", GLOBAL, SHOW_ACCESS) \
     M(SHOW_ACCESS, "", GROUP, ACCESS_MANAGEMENT) \
+    M(IMPERSONATE, "", USER_NAME, ACCESS_MANAGEMENT) \
```
Separating the privilege from the SQL allows us to use the privilege in some other feature in the future. e.g external authentication of the end-user and then IMPERSONATE using some connect time token etc. I have added "EXECUTE AS" an an alias. Thanks!

shiyer7474 committed:
Review comments and implement EXECUTE AS <user> SELECT

shiyer7474 commented:
I will do the documentation changes once we finalize on the syntax and behaviour specifics!

shiyer7474 committed:
Formatting

shiyer7474 commented:
Checking the test failures.

shiyer7474 committed:
Add documentation, test fixes, update query_thread_log.

shiyer7474 commented:
I will add the docs, code review can continue.

Enmk commented:
Hi @vitlibar could you please take a look?

vitlibar commented on src/Parsers/Access/ParserExecuteAsQuery.cpp:
```diff
@@ -0,0 +1,63 @@
+
+    query->targetuser = targetuser;
+
+    /// support 1) EXECUTE AS <user1>  2) EXECUTE AS <user1> SELECT ...
```
It makes sense to support any query after `EXECUTE AS <user>`, not just `SELECT`.

For example,
```
EXECUTE AS <user1> CREATE TABLE ...
```

vitlibar replied:
I added support for that.

vitlibar commented on tests/queries/0_stateless/03252_execute_as.sh:
```diff
@@ -0,0 +1,64 @@
+    "--listen_host=127.1"
+    # we will discover the real port later.
+    "--tcp_port=0"
+    "--shutdown_wait_unfinished=0"
```
It doesn't make sense to start another server in this test.
`access_management` is [already set to 1](https://github.com/ClickHouse/ClickHouse/blob/master/tests/config/users.d/access_management.xml) in our CI.

vitlibar replied:
I rewrote the test and simplified it.

vitlibar commented on src/Interpreters/Access/InterpreterExecuteAsQuery.cpp:
```diff
@@ -0,0 +1,74 @@
+
+    BlockIO return_value = {};
+
+    auto restore_previous_user = [&] () { getContext()->getSessionContext()->switchImpersonateUser(RolesOrUsersSet(current_user_uuid.value())); };
```
Here `restore_previous_user` won't work good enough - because `switchImpersonateUser()` also resets settings and roles. What is better is to clone the query context and you don't need to worry about restoring it afterwards:

```
if (query.subquery)
{
    auto impersonation_context = Context::createCopy(*getContext());
    impersonation_context->makeQueryContext();
    impersonation_context->switchImpersonateUser(...);
    return executeQuery(serializeAST(query.subquery), impersonation_context);
}
else
{
    getContext()->getSessionContext()->switchImpersonateUser();
    return {};
}

p3io commented:
I will resume work on this PR in a week's time, thanks!

vitlibar committed:
Rewrite test.

vitlibar committed:
Add documentation.

vitlibar committed:
Allow any subqueries in EXECUTE AS, not just SELECT.
Use a separated context to handle the subquery in EXECUTE AS <target_user> <subquery>.

vitlibar committed:
Rename column 'auth_user' to 'authenticated_user' in system.query_log

vitlibar committed:
Disable server setting 'allow_impersonate_user' by default.

vitlibar committed:
Rename function authUser() -> authenticatedUser(), keep authUser() as an alias.

vitlibar committed:
Fix docs.

fm4v commented:
@vitlibar @shiyer7474
Negative test when `allow_impersonate_user=0` is missing which is a huge security gap. These cases should be added:
 - Can't run `EXECUTE AS target_user` or `EXECUTE AS target_user subquery;` if server setting disabled even as admin
 - Can't run `EXECUTE AS` even with `IMPERSONATE` grant or as admin
